ratschlab_common/io/sparse_df.py

```diff
@@ -0,0 +1,122 @@
+import pandas as pd
+import numpy as np
+from pandas.api.types import is_sparse
+from scipy import sparse
+import tables
+
+'''
+This module exposes two functions: to_hdf and read_hdf, which usage
+is to read and write sparse Pandas DataFrames into and from hdf file.
+'''
+
+
+def read_hdf(fpath):
+    '''
+    read a hdf file into a sparse Pandas DataFrame that was previously saved with the to_hdf function of this module.
+
+    :param fpath: location of the hdf file
+    :return: an instance of Pandas DataFrame
+    '''
+    f = tables.open_file(fpath, mode='r')
+    sparse_group = f.get_node("/data/sparse")
+    sparse_m = _read_sparse_m_coo(sparse_group)
+    s_col_names = f.get_node("/data/sparse/col_names")
+    non_sparse = pd.read_hdf(fpath, key="/data/non_sparse")
+    sparse = pd.DataFrame.sparse.from_spmatrix(sparse_m, columns=s_col_names)
+    ret_df = pd.concat([sparse, non_sparse], axis=1)
+    f.close()
+    return ret_df
+
+def _read_sparse_m(group):
+    attributes = []
+    for attribute in ('data', 'indices', 'indptr', 'shape'):
+        attributes.append(getattr(group, attribute).read())
+    sparse_m = sparse.csr_matrix(tuple(attributes[:3]), shape=attributes[3])
+    return sparse_m
+
+def _read_sparse_m_coo(group):
+    attributes = []
+    for attribute in ('data', 'row', 'col', 'shape'):
+        attributes.append(getattr(group, attribute).read())
+    t = tuple([attributes[1], attributes[2]])
+    sparse_m = sparse.coo_matrix(tuple([attributes[0], t]), shape=attributes[3])
+    return sparse_m
+
+def _store_sparse_m_coo(m, hdf5, group, complevel=5, complib='blosc', **kwargs):
+    assert (m.__class__ == sparse.coo.coo_matrix), 'm must be a coo matrix'
+    for attribute in ('data', 'row', 'col', 'shape'):
+        arr = np.array(getattr(m, attribute))
+        _store_array_hdf(arr, hdf5, group, attribute,
+                                        complevel=complevel, complib=complib, **kwargs)
+    return hdf5
+
+
+def _store_sparse_m(m, hdf5, group, complevel=5, complib='blosc', **kwargs):
+    assert (m.__class__ == sparse.csr.csr_matrix), 'M must be a csr matrix'
+    for attribute in ('data', 'indices', 'indptr', 'shape'):
+        arr = np.array(getattr(m, attribute))
+        _store_array_hdf(arr, hdf5, group, attribute,
+                                        complevel=complevel, complib=complib, **kwargs)
+    return hdf5
+
+def _store_array_hdf(arr, hdf5, group, name, complevel=5, complib='blosc', **kwargs):
+    if arr.dtype == 'object':
+        arr = arr.astype('U')
+    atom = tables.Atom.from_dtype(arr.dtype)
+    filters = tables.Filters(complevel=complevel, complib=complib, **kwargs)
+    ds = hdf5.create_carray(group, name, atom, arr.shape, filters=filters)
+    ds[:] = arr
+
+def _store_non_sparse_df(df, path, complevel=5, complib='blosc', **kwargs):
+    df.to_hdf(path, key="data/non_sparse",
+                         complevel=complevel, complib=complib, **kwargs)
+
+
+def to_hdf(df, path, **kwargs):
```

Comment:
perhaps it's more natural to have the following signature:
```
def to_hdf(df, path, sparse_hdf_filter_dict=None, non_sparse_hdf_filter_dict=None):
```

(the `None` would need to get replaced by '{}' further in the function)